### PR TITLE
[#6243] Platform: Enable "node-to-client" tls if "node-to-node" tls is enabled.

### PR DESCRIPTION
**Pull request by nishantSharma459**

### CODE REVIEW

**Comment by nishantSharma459** [comment]
The change ensures that the "ClientToNode" TLS  is disabled initially until nd unless "NodeToNode" TLS is enabled.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java [hunk 1]
```diff
@@ -463,6 +474,10 @@ public Result create(UUID customerUUID) {
       LOG.info("Create for {}.", customerUUID);
       // Get the user submitted form data.
       formData = (ObjectNode) request().body().asJson();
+      if (validateEncryption(formData))
+      {
+        return ApiResponse.error(BAD_REQUEST, "Node to node encryption must required to enable for client to node encryption.");
+      }
```
Comment (jaydeepkumara) [inline managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:480]: For now, i just have added validation for creating api let me know if we need to add any other api.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java [hunk 2]
```diff
@@ -124,6 +124,17 @@ public UniverseController(YBClientService service) {
     this.ybService = service;
   }
 
+  private boolean validateEncryption(ObjectNode formData) {
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:127]: Hm... Before this I saw two ways to implement validate functions - without result which throws exceptions if something is wrong, another - which returns true if the arguments is successfully validated. Your implementation can confuse someone as it implements inverted logic - it returns `true` if something is wrong. Actually I prefer standard conventions.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java [hunk 3]
```diff
@@ -124,6 +124,17 @@ public UniverseController(YBClientService service) {
...
+    boolean returnValue = true;
+    for (JsonNode cluster : formData.get("clusters")) {
+      JsonNode nodeToNodeEncryption = cluster.get("userIntent").get("enableNodeToNodeEncrypt");
+      JsonNode clientToNodeEncryption = cluster.get("userIntent").get("enableClientToNodeEncrypt");
+
+      returnValue = returnValue && (!nodeToNodeEncryption.asBoolean() && clientToNodeEncryption.asBoolean());
+    }
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:134]: Despite the previous comment, this one is about the loop - you can exit the loop if the variable becomes false.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java [hunk 4]
```diff
@@ -463,6 +474,10 @@ public Result create(UUID customerUUID) {
       LOG.info("Create for {}.", customerUUID);
       // Get the user submitted form data.
       formData = (ObjectNode) request().body().asJson();
+      if (validateEncryption(formData))
+      {
+        return ApiResponse.error(BAD_REQUEST, "Node to node encryption must required to enable for client to node encryption.");
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:479]: Please rephrase. It is not completely clear what it wants to say.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java (hunk 1 above)
Comment (SergeyPotachev) [reply managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:480]: Please verify that UI processes the possible error correctly. Often it doesn't.

**Review by SergeyPotachev — CHANGES_REQUESTED** [review]

File: managed/ui/src/components/universes/UniverseForm/ClusterFields.js [hunk 5]
```diff
@@ -724,6 +725,12 @@ export default class ClusterFields extends Component {
       updateFormField('primary.enableNodeToNodeEncrypt', event.target.checked);
       updateFormField('async.NodeToNodeEncrypt', event.target.checked);
       this.setState({ enableNodeToNodeEncrypt: event.target.checked });
+    } 
+    
+    // This condition will make sure if enableNodeToNodeEncrypt is false
+    // then enableClientToNodeEncrypt will set to false.
+    if(!event.target.checked) {
+      this.setState({ enableClientToNodeEncrypt: false});
```
Comment (yb-andrew) [inline managed/ui/src/components/universes/UniverseForm/ClusterFields.js:733]: This `setState` call can be moved into the one above. Just check that the `event.target.checked` is false.
```
this.setState({
  enableNodeToNodeEncrypt: event.target.checked,
  enableClientToNodeEncrypt: this.state.enableClientToNodeEncrypt && event.target.checked
});    
```

File: managed/ui/src/components/universes/UniverseForm/ClusterFields.js [hunk 6]
```diff
@@ -1020,6 +1027,16 @@ export default class ClusterFields extends Component {
...
   }
 
+  /**
+   * This method is used to disable the ClientToNodeTLS field initially.
+   * 
+   * @param isFieldReadOnly If true then readonly access.
+   * @param enableNodeToNodeEncrypt NodeToNodeTLS state.
+   */
+  clientToNodeEncryptField(isFieldReadOnly, enableNodeToNodeEncrypt) {
+    return isFieldReadOnly ? isFieldReadOnly : !enableNodeToNodeEncrypt;
```
Comment (yb-andrew) [inline managed/ui/src/components/universes/UniverseForm/ClusterFields.js:1037]: Can change to `return isFieldReadOnly || !enableNodeToNodeEncrypt;`

File: managed/ui/src/components/universes/UniverseForm/ClusterFields.js [hunk 7]
```diff
@@ -1304,7 +1321,7 @@ export default class ClusterFields extends Component {
         <Field
           name={`${clusterType}.enableClientToNodeEncrypt`}
           component={YBToggle}
-          isReadOnly={isFieldReadOnly}
+          isReadOnly={ this.clientToNodeEncryptField(isFieldReadOnly, this.state.enableNodeToNodeEncrypt)}
```
Comment (yb-andrew) [inline managed/ui/src/components/universes/UniverseForm/ClusterFields.js:1324]: Is the intention to only set the field initially? What about when `this.state.enableNodeToNodeEncrypt` changes value? Shouldn't the client-to-node field then become editable?

**Review by yb-andrew — CHANGES_REQUESTED** [review]

File: managed/ui/src/components/universes/UniverseForm/ClusterFields.js (hunk 5 above)
Comment (nishantSharma459) [reply managed/ui/src/components/universes/UniverseForm/ClusterFields.js]: Done.

File: managed/ui/src/components/universes/UniverseForm/ClusterFields.js (hunk 6 above)
Comment (nishantSharma459) [reply managed/ui/src/components/universes/UniverseForm/ClusterFields.js]: Done.

File: managed/ui/src/components/universes/UniverseForm/ClusterFields.js (hunk 7 above)
Comment (nishantSharma459) [reply managed/ui/src/components/universes/UniverseForm/ClusterFields.js:1323]: I've updated the Doc against the same.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java (hunk 2 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:127]: ok got it

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java (hunk 4 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java]: Done

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java (hunk 3 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:135]: Done

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java (hunk 2 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:127]: Done

**Review by yb-andrew — APPROVED** [review]
UI changes look good.

File: managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java (hunk 1 above)
Comment (jaydeepkumara) [reply managed/src/main/java/com/yugabyte/yw/controllers/UniverseController.java:482]: This specific scenario has been handled in the Front end.

**Review by jaydeepkumara — APPROVED** [review]

**Review by SergeyPotachev — APPROVED** [review]
Looks good.

**Pull request merged**